### ISSUE TITLE
Integrate transformer

### ISSUE THREAD

**Issue by juliakreutzer** [joeynmt/joeynmt#6]
Following [The Annotated Transformer](http://nlp.seas.harvard.edu/2018/04/03/attention.html)

**Comment by bastings** [comment]
Work in progress in branch `transformer-joey`

**Comment by bastings** [comment]
Still needed before merging:

- [ ] Beam search for transformer
- [ ] Disable gradient clipping
- [ ] Support batches based on number of tokens, with switch --batch-type=token (default: sentence, as it is now)
- [ ] Nice to have: Src-Trg softmax Weight tying

**Comment by juliakreutzer** [comment]
About gradient clipping: it's [disabled](https://github.com/joeynmt/joeynmt/blob/master/joeynmt/training.py#L389) when not specified.

**Comment by juliakreutzer** [comment]
About weight tying: there's two options (see [Sockeye paper](https://arxiv.org/pdf/1712.05690.pdf) §3.1)
1. src and trg embedding tying: only works with joint vocabulary, already implemented for RNNs in [model.py](https://github.com/joeynmt/joeynmt/blob/joey-transformer/joeynmt/model.py#L229)
2. embedding and output layer tying: output layer input size becomes embedding size - for RNNs the size attentional vector (which is usually much larger than what one would use as embedding size)

**Comment by maxdel** [comment]
Hi,

it might be relevant: https://github.com/pytorch/pytorch/commit/83cec5f3ee1b5eab07cea29cd13da3cf57871f59

**Comment by juliakreutzer** [comment]
Thanks for the pointer, @maxdel! We'll stick to our own implementation for now, for educational purpose and as long as it's not in the stable release. If you find anything in their implementation that we should adapt, let us know :)

**Comment by bastings** [comment]
Closing this issue since transformer has now been merged with master branch.
Improvements can come with separate issues.
